{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# DA-RED-LODs\n",
    "DA-RED-LODs is a biological-based feature selection method whose baseline is supported by Open Targets Platform [1], providing a scored ranking which associates genes with a certain disease using biological evidences from all integrated sources. This score is computed as the harmonic sum of the scores associated to each data type, which in turn are computed as the harmonic sum of the scores of each data source. The calculation of each evidence score is performed taking into account the evidence frequency, the strength of the effect described by it and its confidence [2]. On the basis of considering those related scores, evidences as well as LOD score, proposed method is built.\n",
    "\n",
    "DA-RED-LOD is an iterative algorithm that follows the principle of minimum redundance maximum relevance in biological terms and whose  functional diagram can be seen in the following figure.\n",
    "\n",
    "![**DA-RED-LODs diagram.** Overall flowchart of DA-RED-LOD diagram. It describes the workflow of DA-RED-LOD method](imgs/daredfs.png)\n",
    "\n",
    "\n",
    "The starting point is an empty set of selected genes, $S_G$, a set of possible genes, $G$, a specific or general disease, $D$, and, in the case of dealing with a multiclass problem, a set of pathologies, $D_P$. \n",
    "\n",
    "In the first step, the gene with the highest relevancy is selected, being the first gene added to the selected genes set $S_G$. In the following steps, genes scores are calculated by the following equation, and the gene with maximun score in each step is selected and added to $S_G$. \n",
    "\n",
    "\\begin{equation} \n",
    "SCORE(g,S_G,D,D_P) =  Rel(g,D,D_P) - \\sum_{g_i \\in S_G} \\frac{RED(g,g_i,D)}{|S_G|}\n",
    "\\end{equation}\n",
    "\n",
    "Where $Rel(g,D,D_P)$ is the relevancy of the gene $g$ in relation with the disease $D$ and possible sub-diseases in $D_P$, $RED(g,g_i,D)$ is the redundancy of gene $g$ over gene $g_i$ in relation with a disease $D$. Relevance and redundancy concepts are explained in detail in the following subsections.\n",
    "\n",
    "## Relevancy.\n",
    "\n",
    "The relevance of the genes is calculated taking into account both the LOD score in relation to the study disease and its Disease Association (DA) score. \n",
    "\n",
    "Given a gene $g$, a disease $D$ and an optional set of pathologies $D_p$, the relevance of $g$ with respect to $D$ is calculated by the following equation.\n",
    "\n",
    "\\begin{equation} \n",
    "Rel(g,D,D_P) = DA(g,D) \\times LOD(g,D_P)\n",
    "\\end{equation}\n",
    "\n",
    " Where $DA(g,D)$ is the DA score of the gene $g$ with regard to the disease $D$ and $LOD(g,D_P)$ is the LOD score of the gene $g$ in relation with the sub-pathologies $D_P$ in muticlass problems and relative to $D$ in binary problems. \n",
    " \n",
    "\n",
    " \n",
    "### Disease Association score.\n",
    "Disease Association (DA) score between a gene and a certain disease is based on scores obtained by Open Targets Platform. To obtain DA score, a request to this platform's API is performed, in order to receive related genes with a certain disease. This result is compared with candidates genes available for each problem, so scores for common genes are kept.\n",
    "This value is in range 0 to 1, where 1 means a strong relation between the gene and the diseases, while 0 means no relation. \n",
    "\n",
    "\n",
    "### LOD score.\n",
    "\n",
    "LOD score, or B-statistic, is obtained through limma package. It is a statistical indicator of the probability that a gene is differentially expressed given two health states. It is calculated as the logarithm of the ratio between the probability that the gene is differentially expressed for the disease, and the probability that it is not expressed [3,4]. In multiclass problems LOD score is calculated for each possible healthy state, so the mean of all of them is calculated in order to obtain a single score of each gene.\n",
    "Since this score is given by a logarithm, it can take any real value, so these are normalized in the same range as the DA score, this is, between 0 and 1.\n",
    "\n",
    "## Redundancy.\n",
    "\n",
    "Redundance is calculated based on evidences that join a gene with a certain disease, in which are based DA scores and which are obtained by web queries through KnowSeq package. Redundancy of the gene $g$ over the gene $g_i$ in relation with a certain disease $D$, is calculated by the proportion of $g$ evidences that can be found in $g_i$ evidences, either partially or completely. In other words, the percentage of evidences that $g$ shares with $g_i$. Since DA score is calculated from found evidences, the redundancy of the $g$ over the $g_i$ in relation with $D$ is defined as the proportion of the DA score of $g$ that is explained through $g_i$. This is, following the next equation.\n",
    "\n",
    "\\begin{equation}\n",
    "RED(g,g_i,D) = \\frac{\\textrm{Num. of evidences of $g$ in $g_i$}}{\\textrm{Num. of g evidences}} \\times DA(g,D)\n",
    "\\end{equation}\n",
    "\n",
    "\n",
    "Thus, $ 0 \\leq RED(g,g_i,D) \\leq DA(g,D) \\leq 1$. A redundancy of 0 means that there is no evidences that links $g$ and $D$ which also links $g_i$ and $D$, so both DA scores are independent. Conversely a redundancy of $DA(g,D)$ indicates that all found evidences of $g$ in relation to $D$, also relates $g_i$ to $D$, this is that $g$ DA score can be fully explained through $g_i$.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To show how the DA-RED-LOD feature selector works, an example of use will be computed using two possible types of cancer multiclass data. For this example, kidney or lung cancer can be selected by setting the following variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer.type <- 'lung'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.train <- read.table(paste('data/',cancer.type,'/expression-train-multiclass.csv',sep=''),sep='\\t')\n",
    "labels.train <- read.table(paste('data/',cancer.type,'/labels-train-multiclass.csv',sep=''),sep='\\t')$x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: quantreg\n",
      "\n",
      "Loading required package: SparseM\n",
      "\n",
      "\n",
      "Attaching package: ‘SparseM’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    backsolve\n",
      "\n",
      "\n",
      "Loading required package: mclust\n",
      "\n",
      "Package 'mclust' version 5.4.6\n",
      "Type 'citation(\"mclust\")' for citing this R package in publications.\n",
      "\n",
      "Loading required package: topGO\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: graph\n",
      "\n",
      "\n",
      "Attaching package: ‘graph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stringr’:\n",
      "\n",
      "    boundary\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:httr’:\n",
      "\n",
      "    content\n",
      "\n",
      "\n",
      "Loading required package: GO.db\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:rlist’:\n",
      "\n",
      "    List\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "groupGOTerms: \tGOBPTerm, GOMFTerm, GOCCTerm environments built.\n",
      "\n",
      "\n",
      "Attaching package: ‘topGO’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    members\n",
      "\n",
      "\n",
      "##############################################################################\n",
      "Pathview is an open source software package distributed under GNU General\n",
      "Public License version 3 (GPLv3). Details of GPLv3 is available at\n",
      "http://www.gnu.org/licenses/gpl-3.0.html. Particullary, users are required to\n",
      "formally cite the original Pathview paper (not just mention it) in publications\n",
      "or products. For details, do citation(\"pathview\") within R.\n",
      "\n",
      "The pathview downloads and uses KEGG data. Non-academic uses may require a KEGG\n",
      "license agreement (details at http://www.kegg.jp/kegg/legal.html).\n",
      "##############################################################################\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source('featureSelection.R')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function create for computing the feaute selection process receives an expression matrix with the samples in the rows and the genes in the columns, the samples labels and the disease the user wants to study. *Mode* parameter indicates which FS method is used, options are: da, daRed or daLOD. The user can also select the number of genes to extract using the parameter *maxGenes*. Finally, *returnEvidences* is a boolean parameter which indicates if found evidences of selected genes are returned or not.\n",
    "\n",
    "In this example, 20 genes are going to be selected in relation with selected *cancer.type* diseases by using the three possible methods, and found evidences are going to be returned. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating ranking of biological relevant genes by using DA implementation...\n",
      "Disease Association ranking: GABRA3 ROS1 ZBTB16 HOXC11 SFTPD SCN7A HMGA2 MMP3 ABCA3 SFTPC\n",
      "Obtaining related diseases with the DEGs from targetValidation platform...\n",
      "Evidences acquired successfully!\n"
     ]
    }
   ],
   "source": [
    "featureRankingDA <- featureSelection(data.train,labels.train,disease=cancer.type,mode='da',maxGenes=10,returnEvidences=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating ranking of biological relevant genes by using DA-Red implementation...\n",
      "Disease Association ranking: GABRA3 ROS1 ZBTB16 HOXC11 SFTPD SCN7A HMGA2 MMP3 ABCA3 SFTPC\n",
      "Obtaining related diseases with the DEGs from targetValidation platform...\n",
      "Evidences acquired successfully!\n",
      "Calculating genes scores...\n"
     ]
    }
   ],
   "source": [
    "featureRankingDARED <- featureSelection(data.train,labels.train,disease=cancer.type,mode='daRed',maxGenes=10,returnEvidences=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Disease Association ranking: GABRA3 ROS1 ZBTB16 HOXC11 SFTPD SCN7A HMGA2 MMP3 ABCA3 SFTPC\n",
      "Obtaining related diseases with the DEGs from targetValidation platform...\n",
      "Evidences acquired successfully!\n",
      "Calculating genes LOD scores...\n",
      "More than two classes detected, applying limma multiclass\n",
      "Contrasts: adeno-scc\n",
      " Contrasts: adeno-STN\n",
      " Contrasts: scc-STN\n",
      "Calculating genes scores...\n"
     ]
    }
   ],
   "source": [
    "featureRankingDALOD <- featureSelection(data.train,labels.train,disease=cancer.type,mode='daLOD',maxGenes=10,returnEvidences=TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ranking"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below it is shown the ranking of selected genes and computed scores by each FS mode: DA, DA-RED and DA-RED-LOD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 10 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>featureRankingDA.ranking</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GABRA3</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>ROS1</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>ZBTB16</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>HOXC11</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>SFTPD</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>SCN7A</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>HMGA2</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>MMP3</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>ABCA3</th><td>1</td></tr>\n",
       "\t<tr><th scope=row>SFTPC</th><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 1\n",
       "\\begin{tabular}{r|l}\n",
       "  & featureRankingDA.ranking\\\\\n",
       "  & <dbl>\\\\\n",
       "\\hline\n",
       "\tGABRA3 & 1\\\\\n",
       "\tROS1 & 1\\\\\n",
       "\tZBTB16 & 1\\\\\n",
       "\tHOXC11 & 1\\\\\n",
       "\tSFTPD & 1\\\\\n",
       "\tSCN7A & 1\\\\\n",
       "\tHMGA2 & 1\\\\\n",
       "\tMMP3 & 1\\\\\n",
       "\tABCA3 & 1\\\\\n",
       "\tSFTPC & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 1\n",
       "\n",
       "| <!--/--> | featureRankingDA.ranking &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| GABRA3 | 1 |\n",
       "| ROS1 | 1 |\n",
       "| ZBTB16 | 1 |\n",
       "| HOXC11 | 1 |\n",
       "| SFTPD | 1 |\n",
       "| SCN7A | 1 |\n",
       "| HMGA2 | 1 |\n",
       "| MMP3 | 1 |\n",
       "| ABCA3 | 1 |\n",
       "| SFTPC | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "       featureRankingDA.ranking\n",
       "GABRA3 1                       \n",
       "ROS1   1                       \n",
       "ZBTB16 1                       \n",
       "HOXC11 1                       \n",
       "SFTPD  1                       \n",
       "SCN7A  1                       \n",
       "HMGA2  1                       \n",
       "MMP3   1                       \n",
       "ABCA3  1                       \n",
       "SFTPC  1                       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data.frame(featureRankingDA$ranking)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 10 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>unlist.featureRankingDARED.ranking.</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GABRA3.GABRA3</th><td>1.0000000</td></tr>\n",
       "\t<tr><th scope=row>ABCA3</th><td>1.0000000</td></tr>\n",
       "\t<tr><th scope=row>ROS1</th><td>0.9980284</td></tr>\n",
       "\t<tr><th scope=row>SFTPD</th><td>0.9829535</td></tr>\n",
       "\t<tr><th scope=row>MMP3</th><td>0.9564779</td></tr>\n",
       "\t<tr><th scope=row>APOBEC3B</th><td>0.9535354</td></tr>\n",
       "\t<tr><th scope=row>SFTPC</th><td>0.9470920</td></tr>\n",
       "\t<tr><th scope=row>HMGA2</th><td>0.9384921</td></tr>\n",
       "\t<tr><th scope=row>MMP12</th><td>0.8637369</td></tr>\n",
       "\t<tr><th scope=row>CHRNB4</th><td>0.8460661</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 1\n",
       "\\begin{tabular}{r|l}\n",
       "  & unlist.featureRankingDARED.ranking.\\\\\n",
       "  & <dbl>\\\\\n",
       "\\hline\n",
       "\tGABRA3.GABRA3 & 1.0000000\\\\\n",
       "\tABCA3 & 1.0000000\\\\\n",
       "\tROS1 & 0.9980284\\\\\n",
       "\tSFTPD & 0.9829535\\\\\n",
       "\tMMP3 & 0.9564779\\\\\n",
       "\tAPOBEC3B & 0.9535354\\\\\n",
       "\tSFTPC & 0.9470920\\\\\n",
       "\tHMGA2 & 0.9384921\\\\\n",
       "\tMMP12 & 0.8637369\\\\\n",
       "\tCHRNB4 & 0.8460661\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 1\n",
       "\n",
       "| <!--/--> | unlist.featureRankingDARED.ranking. &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| GABRA3.GABRA3 | 1.0000000 |\n",
       "| ABCA3 | 1.0000000 |\n",
       "| ROS1 | 0.9980284 |\n",
       "| SFTPD | 0.9829535 |\n",
       "| MMP3 | 0.9564779 |\n",
       "| APOBEC3B | 0.9535354 |\n",
       "| SFTPC | 0.9470920 |\n",
       "| HMGA2 | 0.9384921 |\n",
       "| MMP12 | 0.8637369 |\n",
       "| CHRNB4 | 0.8460661 |\n",
       "\n"
      ],
      "text/plain": [
       "              unlist.featureRankingDARED.ranking.\n",
       "GABRA3.GABRA3 1.0000000                          \n",
       "ABCA3         1.0000000                          \n",
       "ROS1          0.9980284                          \n",
       "SFTPD         0.9829535                          \n",
       "MMP3          0.9564779                          \n",
       "APOBEC3B      0.9535354                          \n",
       "SFTPC         0.9470920                          \n",
       "HMGA2         0.9384921                          \n",
       "MMP12         0.8637369                          \n",
       "CHRNB4        0.8460661                          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data.frame(unlist(featureRankingDARED$ranking))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 10 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>unlist.featureRankingDALOD.ranking.</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>TFAP2A</th><td>0.6088828</td></tr>\n",
       "\t<tr><th scope=row>TICRR</th><td>0.4832688</td></tr>\n",
       "\t<tr><th scope=row>CHRNB4</th><td>0.4782881</td></tr>\n",
       "\t<tr><th scope=row>MMP12</th><td>0.4900457</td></tr>\n",
       "\t<tr><th scope=row>E2F7</th><td>0.4110653</td></tr>\n",
       "\t<tr><th scope=row>ABCA3</th><td>0.3775388</td></tr>\n",
       "\t<tr><th scope=row>APOBEC3B</th><td>0.3278171</td></tr>\n",
       "\t<tr><th scope=row>GABRA3</th><td>0.3523289</td></tr>\n",
       "\t<tr><th scope=row>CDC45</th><td>0.3590342</td></tr>\n",
       "\t<tr><th scope=row>SFTPC</th><td>0.2695595</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 1\n",
       "\\begin{tabular}{r|l}\n",
       "  & unlist.featureRankingDALOD.ranking.\\\\\n",
       "  & <dbl>\\\\\n",
       "\\hline\n",
       "\tTFAP2A & 0.6088828\\\\\n",
       "\tTICRR & 0.4832688\\\\\n",
       "\tCHRNB4 & 0.4782881\\\\\n",
       "\tMMP12 & 0.4900457\\\\\n",
       "\tE2F7 & 0.4110653\\\\\n",
       "\tABCA3 & 0.3775388\\\\\n",
       "\tAPOBEC3B & 0.3278171\\\\\n",
       "\tGABRA3 & 0.3523289\\\\\n",
       "\tCDC45 & 0.3590342\\\\\n",
       "\tSFTPC & 0.2695595\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 1\n",
       "\n",
       "| <!--/--> | unlist.featureRankingDALOD.ranking. &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| TFAP2A | 0.6088828 |\n",
       "| TICRR | 0.4832688 |\n",
       "| CHRNB4 | 0.4782881 |\n",
       "| MMP12 | 0.4900457 |\n",
       "| E2F7 | 0.4110653 |\n",
       "| ABCA3 | 0.3775388 |\n",
       "| APOBEC3B | 0.3278171 |\n",
       "| GABRA3 | 0.3523289 |\n",
       "| CDC45 | 0.3590342 |\n",
       "| SFTPC | 0.2695595 |\n",
       "\n"
      ],
      "text/plain": [
       "         unlist.featureRankingDALOD.ranking.\n",
       "TFAP2A   0.6088828                          \n",
       "TICRR    0.4832688                          \n",
       "CHRNB4   0.4782881                          \n",
       "MMP12    0.4900457                          \n",
       "E2F7     0.4110653                          \n",
       "ABCA3    0.3775388                          \n",
       "APOBEC3B 0.3278171                          \n",
       "GABRA3   0.3523289                          \n",
       "CDC45    0.3590342                          \n",
       "SFTPC    0.2695595                          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data.frame(unlist(featureRankingDALOD$ranking))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Interpretability"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check found evidences for the first selected gene by each FS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'GABRA3'"
      ],
      "text/latex": [
       "'GABRA3'"
      ],
      "text/markdown": [
       "'GABRA3'"
      ],
      "text/plain": [
       "[1] \"GABRA3\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$known_drug</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Drug.Name</th><th scope=col>Molecule.Type</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>PROPOFOL</td><td>Small molecule</td></tr>\n",
       "\t<tr><td>MIDAZOLAM</td><td>Small molecule</td></tr>\n",
       "\t<tr><td>SEVOFLURANE</td><td>Small molecule</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$affected_pathway</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Url</th><th scope=col>Reactome.Url</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/28179366</td><td>http://www.reactome.org/PathwayBrowser/#R-HSA-5696395</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/28179366</td><td>http://www.reactome.org/PathwayBrowser/#R-HSA-354192</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$rna_expression</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Url</th><th scope=col>Comparison</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/27699219</td><td>'tumor tissue' vs 'non-malignant tissue'</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/20522636</td><td>'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue'</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$literature</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 4 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Url</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/23617850</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/27081042</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/19048400</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/25089631</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$known\\_drug] A data.frame: 3 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Drug.Name & Molecule.Type\\\\\n",
       " <list> & <list>\\\\\n",
       "\\hline\n",
       "\t PROPOFOL & Small molecule\\\\\n",
       "\t MIDAZOLAM & Small molecule\\\\\n",
       "\t SEVOFLURANE & Small molecule\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$affected\\_pathway] A data.frame: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Url & Reactome.Url\\\\\n",
       " <list> & <list>\\\\\n",
       "\\hline\n",
       "\t http://europepmc.org/abstract/MED/28179366 & http://www.reactome.org/PathwayBrowser/\\#R-HSA-5696395\\\\\n",
       "\t http://europepmc.org/abstract/MED/28179366 & http://www.reactome.org/PathwayBrowser/\\#R-HSA-354192\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$rna\\_expression] A data.frame: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Url & Comparison\\\\\n",
       " <list> & <list>\\\\\n",
       "\\hline\n",
       "\t http://europepmc.org/abstract/MED/27699219 & 'tumor tissue' vs 'non-malignant tissue'\\\\\n",
       "\t http://europepmc.org/abstract/MED/20522636 & 'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue'\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$literature] A data.frame: 4 × 1\n",
       "\\begin{tabular}{l}\n",
       " Url\\\\\n",
       " <list>\\\\\n",
       "\\hline\n",
       "\t http://europepmc.org/abstract/MED/23617850\\\\\n",
       "\t http://europepmc.org/abstract/MED/27081042\\\\\n",
       "\t http://europepmc.org/abstract/MED/19048400\\\\\n",
       "\t http://europepmc.org/abstract/MED/25089631\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$known_drug\n",
       ":   \n",
       "A data.frame: 3 × 2\n",
       "\n",
       "| Drug.Name &lt;list&gt; | Molecule.Type &lt;list&gt; |\n",
       "|---|---|\n",
       "| PROPOFOL | Small molecule |\n",
       "| MIDAZOLAM | Small molecule |\n",
       "| SEVOFLURANE | Small molecule |\n",
       "\n",
       "\n",
       "$affected_pathway\n",
       ":   \n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| Url &lt;list&gt; | Reactome.Url &lt;list&gt; |\n",
       "|---|---|\n",
       "| http://europepmc.org/abstract/MED/28179366 | http://www.reactome.org/PathwayBrowser/#R-HSA-5696395 |\n",
       "| http://europepmc.org/abstract/MED/28179366 | http://www.reactome.org/PathwayBrowser/#R-HSA-354192 |\n",
       "\n",
       "\n",
       "$rna_expression\n",
       ":   \n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| Url &lt;list&gt; | Comparison &lt;list&gt; |\n",
       "|---|---|\n",
       "| http://europepmc.org/abstract/MED/27699219 | 'tumor tissue' vs 'non-malignant tissue' |\n",
       "| http://europepmc.org/abstract/MED/20522636 | 'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue' |\n",
       "\n",
       "\n",
       "$literature\n",
       ":   \n",
       "A data.frame: 4 × 1\n",
       "\n",
       "| Url &lt;list&gt; |\n",
       "|---|\n",
       "| http://europepmc.org/abstract/MED/23617850 |\n",
       "| http://europepmc.org/abstract/MED/27081042 |\n",
       "| http://europepmc.org/abstract/MED/19048400 |\n",
       "| http://europepmc.org/abstract/MED/25089631 |\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$known_drug\n",
       "    Drug.Name  Molecule.Type\n",
       "1    PROPOFOL Small molecule\n",
       "2   MIDAZOLAM Small molecule\n",
       "3 SEVOFLURANE Small molecule\n",
       "\n",
       "$affected_pathway\n",
       "                                         Url\n",
       "1 http://europepmc.org/abstract/MED/28179366\n",
       "2 http://europepmc.org/abstract/MED/28179366\n",
       "                                           Reactome.Url\n",
       "1 http://www.reactome.org/PathwayBrowser/#R-HSA-5696395\n",
       "2  http://www.reactome.org/PathwayBrowser/#R-HSA-354192\n",
       "\n",
       "$rna_expression\n",
       "                                         Url\n",
       "1 http://europepmc.org/abstract/MED/27699219\n",
       "2 http://europepmc.org/abstract/MED/20522636\n",
       "                                                Comparison\n",
       "1                 'tumor tissue' vs 'non-malignant tissue'\n",
       "2 'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue'\n",
       "\n",
       "$literature\n",
       "                                         Url\n",
       "1 http://europepmc.org/abstract/MED/23617850\n",
       "2 http://europepmc.org/abstract/MED/27081042\n",
       "3 http://europepmc.org/abstract/MED/19048400\n",
       "4 http://europepmc.org/abstract/MED/25089631\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(featureRankingDA$evidences)[1]\n",
    "featureRankingDA$evidences[[1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'GABRA3'"
      ],
      "text/latex": [
       "'GABRA3'"
      ],
      "text/markdown": [
       "'GABRA3'"
      ],
      "text/plain": [
       "[1] \"GABRA3\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$known_drug</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Drug.Name</th><th scope=col>Molecule.Type</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>PROPOFOL</td><td>Small molecule</td></tr>\n",
       "\t<tr><td>MIDAZOLAM</td><td>Small molecule</td></tr>\n",
       "\t<tr><td>SEVOFLURANE</td><td>Small molecule</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$affected_pathway</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Url</th><th scope=col>Reactome.Url</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/28179366</td><td>http://www.reactome.org/PathwayBrowser/#R-HSA-5696395</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/28179366</td><td>http://www.reactome.org/PathwayBrowser/#R-HSA-354192</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$rna_expression</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Url</th><th scope=col>Comparison</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/27699219</td><td>'tumor tissue' vs 'non-malignant tissue'</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/20522636</td><td>'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue'</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$literature</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 4 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Url</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/23617850</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/27081042</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/19048400</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/25089631</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$known\\_drug] A data.frame: 3 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Drug.Name & Molecule.Type\\\\\n",
       " <list> & <list>\\\\\n",
       "\\hline\n",
       "\t PROPOFOL & Small molecule\\\\\n",
       "\t MIDAZOLAM & Small molecule\\\\\n",
       "\t SEVOFLURANE & Small molecule\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$affected\\_pathway] A data.frame: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Url & Reactome.Url\\\\\n",
       " <list> & <list>\\\\\n",
       "\\hline\n",
       "\t http://europepmc.org/abstract/MED/28179366 & http://www.reactome.org/PathwayBrowser/\\#R-HSA-5696395\\\\\n",
       "\t http://europepmc.org/abstract/MED/28179366 & http://www.reactome.org/PathwayBrowser/\\#R-HSA-354192\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$rna\\_expression] A data.frame: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Url & Comparison\\\\\n",
       " <list> & <list>\\\\\n",
       "\\hline\n",
       "\t http://europepmc.org/abstract/MED/27699219 & 'tumor tissue' vs 'non-malignant tissue'\\\\\n",
       "\t http://europepmc.org/abstract/MED/20522636 & 'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue'\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$literature] A data.frame: 4 × 1\n",
       "\\begin{tabular}{l}\n",
       " Url\\\\\n",
       " <list>\\\\\n",
       "\\hline\n",
       "\t http://europepmc.org/abstract/MED/23617850\\\\\n",
       "\t http://europepmc.org/abstract/MED/27081042\\\\\n",
       "\t http://europepmc.org/abstract/MED/19048400\\\\\n",
       "\t http://europepmc.org/abstract/MED/25089631\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$known_drug\n",
       ":   \n",
       "A data.frame: 3 × 2\n",
       "\n",
       "| Drug.Name &lt;list&gt; | Molecule.Type &lt;list&gt; |\n",
       "|---|---|\n",
       "| PROPOFOL | Small molecule |\n",
       "| MIDAZOLAM | Small molecule |\n",
       "| SEVOFLURANE | Small molecule |\n",
       "\n",
       "\n",
       "$affected_pathway\n",
       ":   \n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| Url &lt;list&gt; | Reactome.Url &lt;list&gt; |\n",
       "|---|---|\n",
       "| http://europepmc.org/abstract/MED/28179366 | http://www.reactome.org/PathwayBrowser/#R-HSA-5696395 |\n",
       "| http://europepmc.org/abstract/MED/28179366 | http://www.reactome.org/PathwayBrowser/#R-HSA-354192 |\n",
       "\n",
       "\n",
       "$rna_expression\n",
       ":   \n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| Url &lt;list&gt; | Comparison &lt;list&gt; |\n",
       "|---|---|\n",
       "| http://europepmc.org/abstract/MED/27699219 | 'tumor tissue' vs 'non-malignant tissue' |\n",
       "| http://europepmc.org/abstract/MED/20522636 | 'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue' |\n",
       "\n",
       "\n",
       "$literature\n",
       ":   \n",
       "A data.frame: 4 × 1\n",
       "\n",
       "| Url &lt;list&gt; |\n",
       "|---|\n",
       "| http://europepmc.org/abstract/MED/23617850 |\n",
       "| http://europepmc.org/abstract/MED/27081042 |\n",
       "| http://europepmc.org/abstract/MED/19048400 |\n",
       "| http://europepmc.org/abstract/MED/25089631 |\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$known_drug\n",
       "    Drug.Name  Molecule.Type\n",
       "1    PROPOFOL Small molecule\n",
       "2   MIDAZOLAM Small molecule\n",
       "3 SEVOFLURANE Small molecule\n",
       "\n",
       "$affected_pathway\n",
       "                                         Url\n",
       "1 http://europepmc.org/abstract/MED/28179366\n",
       "2 http://europepmc.org/abstract/MED/28179366\n",
       "                                           Reactome.Url\n",
       "1 http://www.reactome.org/PathwayBrowser/#R-HSA-5696395\n",
       "2  http://www.reactome.org/PathwayBrowser/#R-HSA-354192\n",
       "\n",
       "$rna_expression\n",
       "                                         Url\n",
       "1 http://europepmc.org/abstract/MED/27699219\n",
       "2 http://europepmc.org/abstract/MED/20522636\n",
       "                                                Comparison\n",
       "1                 'tumor tissue' vs 'non-malignant tissue'\n",
       "2 'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue'\n",
       "\n",
       "$literature\n",
       "                                         Url\n",
       "1 http://europepmc.org/abstract/MED/23617850\n",
       "2 http://europepmc.org/abstract/MED/27081042\n",
       "3 http://europepmc.org/abstract/MED/19048400\n",
       "4 http://europepmc.org/abstract/MED/25089631\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(featureRankingDARED$evidences)[1]\n",
    "featureRankingDARED$evidences[[1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'TFAP2A'"
      ],
      "text/latex": [
       "'TFAP2A'"
      ],
      "text/markdown": [
       "'TFAP2A'"
      ],
      "text/plain": [
       "[1] \"TFAP2A\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$affected_pathway</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Url</th><th scope=col>Reactome.Url</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/28179366</td><td>http://www.reactome.org/PathwayBrowser/#R-HSA-3134975</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/28179366</td><td>http://www.reactome.org/PathwayBrowser/#R-HSA-6802949</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/28179366</td><td>http://www.reactome.org/PathwayBrowser/#R-HSA-202403</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$animal_model</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 1 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Is.Associated</th><th scope=col>Specie</th></tr>\n",
       "\t<tr><th scope=col>&lt;named list&gt;</th><th scope=col>&lt;named list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>FALSE</td><td>mouse</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$rna_expression</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 7 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Url</th><th scope=col>Comparison</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/27699219</td><td>'tumor tissue' vs 'non-malignant tissue'</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/20522636</td><td>'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue'</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/20878980</td><td>'non-small cell lung cancer' vs 'normal'</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/20522636</td><td>'lung cancer' vs 'normal' in 'lung; Formalin-fixed paraffin-embedded tissue'</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/20802022</td><td>'primary tumor' vs 'adjacent normal tissue'</td></tr>\n",
       "\t<tr><td>*</td><td>'squamous cell carcinoma' vs 'normal tissue adjacent to squamous cell carcinoma'</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/15653641</td><td>'lung adenocarcinoma' vs 'normal'</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$literature</dt>\n",
       "\t\t<dd><table>\n",
       "<caption>A data.frame: 20 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Url</th></tr>\n",
       "\t<tr><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/16145912</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/25050743</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/25294805</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/25481043</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/29100274</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/15864740</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/17237224</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/31337972</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/32015686</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/28749936</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/26672764</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/31866302</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/23869238</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/22468095</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/1953710</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/30177858</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/28387721</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/26968737</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/29796116</td></tr>\n",
       "\t<tr><td>http://europepmc.org/abstract/MED/26402252</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$affected\\_pathway] A data.frame: 3 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Url & Reactome.Url\\\\\n",
       " <list> & <list>\\\\\n",
       "\\hline\n",
       "\t http://europepmc.org/abstract/MED/28179366 & http://www.reactome.org/PathwayBrowser/\\#R-HSA-3134975\\\\\n",
       "\t http://europepmc.org/abstract/MED/28179366 & http://www.reactome.org/PathwayBrowser/\\#R-HSA-6802949\\\\\n",
       "\t http://europepmc.org/abstract/MED/28179366 & http://www.reactome.org/PathwayBrowser/\\#R-HSA-202403\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$animal\\_model] A data.frame: 1 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Is.Associated & Specie\\\\\n",
       " <named list> & <named list>\\\\\n",
       "\\hline\n",
       "\t FALSE & mouse\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$rna\\_expression] A data.frame: 7 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Url & Comparison\\\\\n",
       " <list> & <list>\\\\\n",
       "\\hline\n",
       "\t http://europepmc.org/abstract/MED/27699219 & 'tumor tissue' vs 'non-malignant tissue'\\\\\n",
       "\t http://europepmc.org/abstract/MED/20522636 & 'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue'\\\\\n",
       "\t http://europepmc.org/abstract/MED/20878980 & 'non-small cell lung cancer' vs 'normal'\\\\\n",
       "\t http://europepmc.org/abstract/MED/20522636 & 'lung cancer' vs 'normal' in 'lung; Formalin-fixed paraffin-embedded tissue'\\\\\n",
       "\t http://europepmc.org/abstract/MED/20802022 & 'primary tumor' vs 'adjacent normal tissue'\\\\\n",
       "\t * & 'squamous cell carcinoma' vs 'normal tissue adjacent to squamous cell carcinoma'\\\\\n",
       "\t http://europepmc.org/abstract/MED/15653641 & 'lung adenocarcinoma' vs 'normal'\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$literature] A data.frame: 20 × 1\n",
       "\\begin{tabular}{l}\n",
       " Url\\\\\n",
       " <list>\\\\\n",
       "\\hline\n",
       "\t http://europepmc.org/abstract/MED/16145912\\\\\n",
       "\t http://europepmc.org/abstract/MED/25050743\\\\\n",
       "\t http://europepmc.org/abstract/MED/25294805\\\\\n",
       "\t http://europepmc.org/abstract/MED/25481043\\\\\n",
       "\t http://europepmc.org/abstract/MED/29100274\\\\\n",
       "\t http://europepmc.org/abstract/MED/15864740\\\\\n",
       "\t http://europepmc.org/abstract/MED/17237224\\\\\n",
       "\t http://europepmc.org/abstract/MED/31337972\\\\\n",
       "\t http://europepmc.org/abstract/MED/32015686\\\\\n",
       "\t http://europepmc.org/abstract/MED/28749936\\\\\n",
       "\t http://europepmc.org/abstract/MED/26672764\\\\\n",
       "\t http://europepmc.org/abstract/MED/31866302\\\\\n",
       "\t http://europepmc.org/abstract/MED/23869238\\\\\n",
       "\t http://europepmc.org/abstract/MED/22468095\\\\\n",
       "\t http://europepmc.org/abstract/MED/1953710\\\\\n",
       "\t http://europepmc.org/abstract/MED/30177858\\\\\n",
       "\t http://europepmc.org/abstract/MED/28387721\\\\\n",
       "\t http://europepmc.org/abstract/MED/26968737\\\\\n",
       "\t http://europepmc.org/abstract/MED/29796116\\\\\n",
       "\t http://europepmc.org/abstract/MED/26402252\\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$affected_pathway\n",
       ":   \n",
       "A data.frame: 3 × 2\n",
       "\n",
       "| Url &lt;list&gt; | Reactome.Url &lt;list&gt; |\n",
       "|---|---|\n",
       "| http://europepmc.org/abstract/MED/28179366 | http://www.reactome.org/PathwayBrowser/#R-HSA-3134975 |\n",
       "| http://europepmc.org/abstract/MED/28179366 | http://www.reactome.org/PathwayBrowser/#R-HSA-6802949 |\n",
       "| http://europepmc.org/abstract/MED/28179366 | http://www.reactome.org/PathwayBrowser/#R-HSA-202403 |\n",
       "\n",
       "\n",
       "$animal_model\n",
       ":   \n",
       "A data.frame: 1 × 2\n",
       "\n",
       "| Is.Associated &lt;named list&gt; | Specie &lt;named list&gt; |\n",
       "|---|---|\n",
       "| FALSE | mouse |\n",
       "\n",
       "\n",
       "$rna_expression\n",
       ":   \n",
       "A data.frame: 7 × 2\n",
       "\n",
       "| Url &lt;list&gt; | Comparison &lt;list&gt; |\n",
       "|---|---|\n",
       "| http://europepmc.org/abstract/MED/27699219 | 'tumor tissue' vs 'non-malignant tissue' |\n",
       "| http://europepmc.org/abstract/MED/20522636 | 'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue' |\n",
       "| http://europepmc.org/abstract/MED/20878980 | 'non-small cell lung cancer' vs 'normal' |\n",
       "| http://europepmc.org/abstract/MED/20522636 | 'lung cancer' vs 'normal' in 'lung; Formalin-fixed paraffin-embedded tissue' |\n",
       "| http://europepmc.org/abstract/MED/20802022 | 'primary tumor' vs 'adjacent normal tissue' |\n",
       "| * | 'squamous cell carcinoma' vs 'normal tissue adjacent to squamous cell carcinoma' |\n",
       "| http://europepmc.org/abstract/MED/15653641 | 'lung adenocarcinoma' vs 'normal' |\n",
       "\n",
       "\n",
       "$literature\n",
       ":   \n",
       "A data.frame: 20 × 1\n",
       "\n",
       "| Url &lt;list&gt; |\n",
       "|---|\n",
       "| http://europepmc.org/abstract/MED/16145912 |\n",
       "| http://europepmc.org/abstract/MED/25050743 |\n",
       "| http://europepmc.org/abstract/MED/25294805 |\n",
       "| http://europepmc.org/abstract/MED/25481043 |\n",
       "| http://europepmc.org/abstract/MED/29100274 |\n",
       "| http://europepmc.org/abstract/MED/15864740 |\n",
       "| http://europepmc.org/abstract/MED/17237224 |\n",
       "| http://europepmc.org/abstract/MED/31337972 |\n",
       "| http://europepmc.org/abstract/MED/32015686 |\n",
       "| http://europepmc.org/abstract/MED/28749936 |\n",
       "| http://europepmc.org/abstract/MED/26672764 |\n",
       "| http://europepmc.org/abstract/MED/31866302 |\n",
       "| http://europepmc.org/abstract/MED/23869238 |\n",
       "| http://europepmc.org/abstract/MED/22468095 |\n",
       "| http://europepmc.org/abstract/MED/1953710 |\n",
       "| http://europepmc.org/abstract/MED/30177858 |\n",
       "| http://europepmc.org/abstract/MED/28387721 |\n",
       "| http://europepmc.org/abstract/MED/26968737 |\n",
       "| http://europepmc.org/abstract/MED/29796116 |\n",
       "| http://europepmc.org/abstract/MED/26402252 |\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$affected_pathway\n",
       "                                         Url\n",
       "1 http://europepmc.org/abstract/MED/28179366\n",
       "2 http://europepmc.org/abstract/MED/28179366\n",
       "3 http://europepmc.org/abstract/MED/28179366\n",
       "                                           Reactome.Url\n",
       "1 http://www.reactome.org/PathwayBrowser/#R-HSA-3134975\n",
       "2 http://www.reactome.org/PathwayBrowser/#R-HSA-6802949\n",
       "3  http://www.reactome.org/PathwayBrowser/#R-HSA-202403\n",
       "\n",
       "$animal_model\n",
       "  Is.Associated Specie\n",
       "1         FALSE  mouse\n",
       "\n",
       "$rna_expression\n",
       "                                         Url\n",
       "1 http://europepmc.org/abstract/MED/27699219\n",
       "2 http://europepmc.org/abstract/MED/20522636\n",
       "3 http://europepmc.org/abstract/MED/20878980\n",
       "4 http://europepmc.org/abstract/MED/20522636\n",
       "5 http://europepmc.org/abstract/MED/20802022\n",
       "6                                          *\n",
       "7 http://europepmc.org/abstract/MED/15653641\n",
       "                                                                        Comparison\n",
       "1                                         'tumor tissue' vs 'non-malignant tissue'\n",
       "2                         'lung cancer' vs 'normal' in 'lung; Fresh-frozen tissue'\n",
       "3                                         'non-small cell lung cancer' vs 'normal'\n",
       "4     'lung cancer' vs 'normal' in 'lung; Formalin-fixed paraffin-embedded tissue'\n",
       "5                                      'primary tumor' vs 'adjacent normal tissue'\n",
       "6 'squamous cell carcinoma' vs 'normal tissue adjacent to squamous cell carcinoma'\n",
       "7                                                'lung adenocarcinoma' vs 'normal'\n",
       "\n",
       "$literature\n",
       "                                          Url\n",
       "1  http://europepmc.org/abstract/MED/16145912\n",
       "2  http://europepmc.org/abstract/MED/25050743\n",
       "3  http://europepmc.org/abstract/MED/25294805\n",
       "4  http://europepmc.org/abstract/MED/25481043\n",
       "5  http://europepmc.org/abstract/MED/29100274\n",
       "6  http://europepmc.org/abstract/MED/15864740\n",
       "7  http://europepmc.org/abstract/MED/17237224\n",
       "8  http://europepmc.org/abstract/MED/31337972\n",
       "9  http://europepmc.org/abstract/MED/32015686\n",
       "10 http://europepmc.org/abstract/MED/28749936\n",
       "11 http://europepmc.org/abstract/MED/26672764\n",
       "12 http://europepmc.org/abstract/MED/31866302\n",
       "13 http://europepmc.org/abstract/MED/23869238\n",
       "14 http://europepmc.org/abstract/MED/22468095\n",
       "15  http://europepmc.org/abstract/MED/1953710\n",
       "16 http://europepmc.org/abstract/MED/30177858\n",
       "17 http://europepmc.org/abstract/MED/28387721\n",
       "18 http://europepmc.org/abstract/MED/26968737\n",
       "19 http://europepmc.org/abstract/MED/29796116\n",
       "20 http://europepmc.org/abstract/MED/26402252\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(featureRankingDALOD$evidences)[1]\n",
    "featureRankingDALOD$evidences[[1]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:2:90: unexpected symbol\n1: data.test <- t(read.table(paste('data/',cancer.type,'/expression-test-multiclass.csv',sep=''),sep='\\t'))\n2: labels.test <- read.table(paste('data/',cancer.type,'/labels-test-multiclass.csv',sep='')sep\n                                                                                            ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:2:90: unexpected symbol\n1: data.test <- t(read.table(paste('data/',cancer.type,'/expression-test-multiclass.csv',sep=''),sep='\\t'))\n2: labels.test <- read.table(paste('data/',cancer.type,'/labels-test-multiclass.csv',sep='')sep\n                                                                                            ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "data.test <- t(read.table(paste('data/',cancer.type,'/expression-test-multiclass.csv',sep=''),sep='\\t'))\n",
    "labels.test <- read.table(paste('data/',cancer.type,'/labels-test-multiclass.csv',sep='')sep='\\t')$x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "test.results.da <- knn_test(data.train,labels.train,data.test,labels.test,names(featureRankingDA$ranking))\n",
    "test.results.daRed <- knn_test(data.train,labels.train,data.test,labels.test,names(featureRankingDARED$ranking))\n",
    "test.results.daLOD <- knn_test(data.train,labels.train,data.test,labels.test,names(featureRankingDALOD$ranking))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot\n",
    "data <- data.frame('da'=test.results.da$accVector,'daRed'=test.results.daRed$accVector,\n",
    "                   'daLod'=test.results.daLOD$accVector,'ngenes'=seq(1,length(da.results$accVector)))\n",
    "ggplot(data = data,x=ngenes) +\n",
    "geom_line(aes(x=ngenes,y=da,colour = 'da')) +\n",
    "geom_line(aes(x=ngenes,y=daRed,colour = 'daRed')) +\n",
    "geom_line(aes(x=ngenes,y=daLod,colour = 'daLod')) +\n",
    "labs (y = 'Accuracy',x='Numer of genes') + ggtitle ('Accuracy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    1. Carvalho-Silva, D., Pierleoni, A., Pignatelli, M., Ong, C., Fumis, L., Karamanis, N., ... & Miranda, A. (2019). Open Targets Platform: new developments and updates two years on. Nucleic acids research, 47(D1), D1056-D1065.\n",
    "    2. Koscielny, G., An, P., Carvalho-Silva, D., Cham, J. A., Fumis, L., Gasparyan, R., ... & Pierleoni, A. (2017). Open Targets: a platform for therapeutic target identification and validation. Nucleic acids research, 45(D1), D985-D994.\n",
    "    3. Ritchie, M. E., Phipson, B., Wu, D. I., Hu, Y., Law, C. W., Shi, W., & Smyth, G. K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic acids research, 43(7), e47-e47\n",
    "    4. Nyholt, D. R. (2000). All LODs are not created equal. The American Journal of Human Genetics, 67(2), 282-288."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
